File: Hard/23.cpp
/** bigO = n lgk
 * works by divide and conquer similar to merge sort. Hardest part:
 * how do you write a while and for loop so that it merges the correct lists
 * in the right order!!!
 * 
 * start using nullptr over NULL
 * 
 * 
 * 
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {

        if(lists.empty())
        {
            return nullptr;
        }

        int k = lists.size();

        int distance = 1;

        while(distance < k)
        {
            cout << "k:" << k << endl;
            for(int i = 0; i < k - distance; i = i + (distance * 2))
            {
                cout << "i: " << i << endl;
                lists[i] = merge(lists[i], lists[i+distance]);
            }
            distance = distance * 2;
        }

        return lists[0];
      

    }

    ListNode* merge(ListNode* list1, ListNode* list2)
    {
        ListNode* it1 = list1;
        ListNode* it2 = list2;

        ListNode* head = new ListNode(0);
        ListNode* current = head;

        if(it1 == NULL)
        {
            return it2;
        }

        if(it2 == NULL)
        {
            return it1;
        }
        
        while(it1 != NULL && it2 != NULL)
        {
            if(it1->val <= it2->val)
            {
                current->next = it1;
                it1 = it1->next;
            }
            else
            {
                current->next = it2;
                it2 = it2->next;
            }
            current = current->next;
        }

        if(it1 == NULL)
        {
            current->next = it2;
        }
        else
        {
            current->next = it1;
        }
        return head->next;
    }
};

File: Hard/23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if len(lists) == 0:
            return None
        distance = 1
        k = len(lists)
        while distance < k:
            #print(f'Distance is: {distance}')
            for i in range(0, k - distance, distance * 2):
                #print(f'i = {i}')
                lists[i] = self.merge(lists[i], lists[i + distance])
            distance = distance * 2
        return lists[0]
    
    def merge(self, l1: ListNode, l2: ListNode):
        newListHead = ListNode()
        newList = newListHead
        attachNode = None
        current1 = l1
        current2 = l2
        while current1 != None and current2 != None:
            if current1.val <= current2.val:
                attachNode = current1
                current1 = current1.next
            else:
                attachNode = current2
                current2 = current2.next
            newList.next = attachNode
            newList = newList.next
        if current1 == None:
            newList.next = current2
        elif current2 == None:
            newList.next = current1
        return newListHead.next
